neuralpp/inference/graphical_model/representation/factor/symbolic_factor.py

```diff
@@ -0,0 +1,133 @@
+from typing import List, Any, Optional, Type, Callable, Dict
+import operator
+
+from neuralpp.inference.graphical_model.representation.factor.atomic_factor import AtomicFactor
+from neuralpp.inference.graphical_model.variable.discrete_variable import DiscreteVariable
+from neuralpp.inference.graphical_model.variable.variable import Variable
+from neuralpp.inference.graphical_model.variable.integer_variable import IntegerVariable
+from neuralpp.symbolic.basic_expression import BasicConstant, BasicVariable, BasicFunctionApplication
+from neuralpp.symbolic.expression import Expression, FunctionApplication
+from neuralpp.symbolic.sympy_interpreter import SymPyInterpreter
+from neuralpp.util.util import union, join
+
+class SymbolicFactor(AtomicFactor):
+    def __init__(self, variables: List[Variable], expression: Expression):
+        super().__init__(variables)
+        self.expression = expression
+
+    def new_instance(self, variables: List[Variable], expression: Expression):
+        return type(self)(variables, expression)
+
+    def assignments(self):
+        return DiscreteVariable.assignments_product(self.variables)
+
+    def _variable_type(self, variable: Variable) -> type:
+        if (isinstance(variable, IntegerVariable)):
+            return int
+        else:
+            return float
+
+    def _dict_to_context(self, assignment_dict: dict) -> Expression:
+        result = BasicConstant(True)
+        type_dict = {}
+        for k, v in assignment_dict.items():
+            symbol = BasicVariable(k.name, self._variable_type(v))
+            eq_operator = BasicConstant(operator.eq, Callable[[type(v), type(v)], bool])
```

Comment:
> Overloaded operator here too.

I'm not quite sure how I can overload `eq` here, since I would need something to contain the expression `symbol = v`. I don't think I can contain that in `BasicVariable` or `BasicConstant`